### ISSUE TITLE
idea: Context.DrawSVG() that draws a parsed SVG

### ISSUE THREAD

**Issue by supertxtnet** [tdewolff/canvas#235]
Canvas is already a super capable project with the drawing/painting primitives, the variety of renderers, and the DrawImage/DrawPath functions. Given the canvases already has state push/popping, it would be a great addition if there was a DrawSVG() that can take a parsed SVG representation and apply the various paths, transformations, etc. and return the canvas to the previous state.

If this were to be considered as part of canvas it might be worth pointing out that there is already a Go SVG parsing and rendering library [oksvg](https://github.com/srwiley/oksvg). It's however tightly coupled to its own rasterx library with a limited set of rendering back-ends. Notably, it doesn't support PDF, nor GioUI (or OpenGL) for putting SVG renderings into UI's.

There is however a [fork](https://github.com/benoitkugler/oksvg) of that project that abstracts a Driver interface for the renderer with an experimental PDF driver. It's not nearly as fully-features as the one here. I've managed to write a Driver that hooks it into Canvas with some decent success with only a few hours of work, which I hope to contribute at some point once I manage to get it into a stable state. The fork has also drifted in terms of bug fixes from the original project.

I think that perhaps it makes sense to combine the energy of these projects into one place. Canvas already has SVG capabilities for rendering, why not also some support for parsing, and drawing as well? Canvas already seems to be far more active than oksvg, and especially its fork.

Thank you for your consideration.

**Comment by tdewolff** [comment]
Thanks for the kind words and an excellent idea. One comment though, there is no need to work with a state or a `Context`, rather you can draw to a renderer directly by using the `Renderer` interface. Rather than the `Canvas.RenderTo` function, it would be a function signature of `RenderSVG(r Renderer, svg io.Reader) error` which would parse the SVG and render the elements to the `Renderer`.

I already have an XML parser at https://github.com/tdewolff/parse/tree/master/xml and it is fairly easy to parse some basic SVGs. The problem I foresee is that it is quite an undertaking to support rendering entire SVGs since there is a lot of functionality to cover (think of CSS, `<defs>`, etc.). The subset that oksvg support though is quite easy to achieve: viewbox, rect/line/circle/polyline/polygon/path, desc, g, title, defs/use, linear/radial gradient. Maybe if I have some time free soon I can sketch up a prototype!

**Comment by supertxtnet** [comment]
I'm impressed with the number of svg files that are working with oksvg, so perhaps that subset is enough to get a broad SVG support working. I would recommend instead of RenderSVG() taking a reader with the SVG, that instead there can be a parsed representation of the SVG in memory, so that the parse step can be done once and it can be rendered multiple times. This is particularly useful with Gio, which tends to rerender things frequently.

**Comment by tdewolff** [comment]
I've pushed a simple implementation of parsing SVGs to a `Canvas`. It supports basic draw tags: circle, ellipse, path, polygon, polyline, rect. And it support basic representation attributes: fill, stroke, stroke-width, stroke-dashoffset, stroke-dasharray, stroke-linecap, stroke-linejoin. It parses colors and dimensions and takes care of SVG dimensions and the viewBox. It also keeps track of a style stack and pushes/pops from the stack when entering/exiting groups or other elements.

It misses functionality for the `transform` attribute, for linear/radial gradients, and for defs/use. Is there anything else required according to you? We'll go and add functionality as needed.

**Comment by supertxtnet** [comment]
This is fantastic, thank you!

I hope to give this a try shortly with some of the drawings I've been testing to see if my new oksvg/canvas integration can be obsoleted before I even really shared it widely. https://supertxt.net/git/svgcanvas

From the samples of programs that I'm hoping to integrate I'd say that defs are pretty common, and style can be found outside of the defs at the top. Transform is less common, but would be great to have. Gradients aren't so common from tools like graphviz/svgbob, but can be found sometimes in beautiful icons (e.g. libreoffice) with linear being much more common that radial.

Text would be really nice to have too, which is currently a limitation in oksvg leaving diagrams without their labels. :(

Here's to hoping that someday we'll have great graphics capabilities in native Go apps that might put them closer to being on par with web apps.

**Comment by supertxtnet** [comment]
The new ParseSVG() is working well so far in my tests.

When I renderTo the canvas, how to I translate it so that the anchor point of it is on the top-left instead of bottom-left? I'd prefer that the drawings be clipped on the bottom than the top.

**Comment by supertxtnet** [comment]
I've created #240 to add lines, styles, and text support to ParseSVG, which makes my tests work much better.

**Comment by supertxtnet** [comment]
I've added #241 to add markers, better font handling, and some fixes. I ran that through a series of random SVG's on my system the current support has a very high success (no errors) rate at parsing and rendering the images. The success rate at rendering something that appears close to correct is a bit lower, possibly in the 75% range.

Once that PR is merged, I wonder if we can call this idea done and the issue can be closed. Fixing the bugs and adding smaller features can be captured in new issues.

@tdewolff does this sound reasonable?

**Comment by supertxtnet** [comment]
@tdewolff is there a place in the Context where one can push/pop extra information like fonts, and the marker canvases? I couldn't seem to find anything in there for arbitrary data.

**Comment by tdewolff** [comment]
Yeah I figured we needed something to keep extra state. I've added it as `svgState` to the `svgParser`. Hope that helps!

**Comment by tdewolff** [comment]
See https://github.com/tdewolff/canvas/commit/2929017693e7aa71a644dc236e04e3cfc75579d5 for initial marker support, it's missing orientation and proper sizing.

**Comment by tdewolff** [comment]
Orientation and sizing should be working correctly now. Fixed the bug when `Coords` and `CoordDirections` returned a different amount of points. Also fixed bugs related to font selection from system. Considering this issue as done, please comment / reopen / open a new issue for bugs or new feature requests! Thanks for all the effort @supertxtnet ! Great team effort :-)

**Comment by supertxtnet** [comment]
Thanks so much @tdewolff. This is working well on my end now, which is a graphical unix shell with support for viewing images, such as svg. :-)

I think that pull request #251 can be closed now.
